contracts/starknet/space/space.cairo

```diff
@@ -0,0 +1,256 @@
+        tempvar pedersen_ptr = pedersen_ptr
+        tempvar voting_power = voting_power
+    else:
+        if choice == Choice.AGAINST:
+            let (against) = power_against.read(proposal_id)
+            # TODO: what should we do about the carry
+            let (power, carry) = uint256_add(against, voting_power)
+            power_against.write(proposal_id, power)
+            tempvar range_check_ptr = range_check_ptr
+            tempvar syscall_ptr = syscall_ptr
+            tempvar pedersen_ptr = pedersen_ptr
+            tempvar voting_power = voting_power
+        else:
+            if choice == Choice.ABSTAIN:
+                let (_abstain) = power_abstain.read(proposal_id)
+                # TODO: what should we do about the carry
+                let (power, carry) = uint256_add(_abstain, voting_power)
+                power_abstain.write(proposal_id, power)
+                tempvar range_check_ptr = range_check_ptr
+                tempvar syscall_ptr = syscall_ptr
+                tempvar pedersen_ptr = pedersen_ptr
+                tempvar voting_power = voting_power
+            else:
+                # choice is not a valid choice ?!
+                return ()
+            end
+        end
+    end
+
+    let vote = Vote(choice=choice, voting_power=voting_power)
+    vote_registry.write(proposal_id, voter_address, vote)
+
+    return ()
+end
+
+# TODO: execution_hash should be of type Hash and metadata_uri of type felt* (string)
+@external
+func propose{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr : felt}(
+        proposer_address : EthAddress, execution_hash : felt, metadata_uri : felt) -> ():
+    alloc_locals
+
+    # Verify that the caller is the authenticator contract.
+    authenticator_only()
+
+    let (current_block) = get_block_number()
+    let (delay) = voting_delay.read()
+    let (duration) = voting_period.read()
+
+    # Define start_block and end_block based on current block, delay and duration variables.
+    let start_block = current_block + delay
+    let end_block = start_block + duration
+
+    # Get the voting power of the proposer
+    let (strategy_contract) = voting_strategy.read()
+    let (local params : felt*) = alloc()
+    let (voting_power) = IVotingStrategy.get_voting_power(
+        contract_address=strategy_contract,
+        address=proposer_address,
+        at=current_block,
+        params_len=0,
+        params=params)
+
+    # Verify that the proposer has enough voting power to trigger a proposal
+    let (threshold) = proposal_threshold.read()
+    let (lower) = uint256_lt(voting_power, threshold)
+    if lower == 1:
+        # Not enough voting power to create a proposal
+        assert 1 = 0
+    end
+
+    # Create the proposal and its proposal id
+    let proposal = Proposal(execution_hash, metadata_uri, start_block, end_block)
+    let (proposal_id) = next_proposal_nonce.read()
+
+    # Store the proposal
+    proposal_registry.write(proposal_id, proposal)
```

Comment:
Can you emit event here with all the info + proposer_address

Reply:
[5997c7b](https://github.com/snapshot-labs/sx-core/pull/47/commits/5997c7b45cce5cac00f09c5bd84fe6fc4b142120) :D